cryptoki/src/slot/token_info.rs

```diff
@@ -0,0 +1,553 @@
+// Copyright 2021 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+//! PKCS11 Token info and associated flags
+
+use crate::error::{Error, Result};
+use crate::string_from_blank_padded;
+use crate::types::convert_utc_time;
+use crate::types::{UtcTime, Version};
+use bitflags::bitflags;
+use cryptoki_sys::*;
+use std::convert::TryFrom;
+use std::fmt::Debug;
+
+bitflags! {
+    /// Collection of flags defined for [`CK_TOKEN_INFO`]
+    struct TokenInfoFlags: CK_FLAGS {
+        const RNG = CKF_RNG;
+        const WRITE_PROTECTED = CKF_WRITE_PROTECTED;
+        const LOGIN_REQUIRED = CKF_LOGIN_REQUIRED;
+        const USER_PIN_INITIALIZED = CKF_USER_PIN_INITIALIZED;
+        const RESTORE_KEY_NOT_NEEDED = CKF_RESTORE_KEY_NOT_NEEDED;
+        const CLOCK_ON_TOKEN = CKF_CLOCK_ON_TOKEN;
+        const PROTECTED_AUTHENTICATION_PATH = CKF_PROTECTED_AUTHENTICATION_PATH;
+        const DUAL_CRYPTO_OPERATIONS = CKF_DUAL_CRYPTO_OPERATIONS;
+        const TOKEN_INITIALIZED = CKF_TOKEN_INITIALIZED;
+        const SECONDARY_AUTHENTICATION = CKF_SECONDARY_AUTHENTICATION;
+        const USER_PIN_COUNT_LOW = CKF_USER_PIN_COUNT_LOW;
+        const USER_PIN_FINAL_TRY = CKF_USER_PIN_FINAL_TRY;
+        const USER_PIN_LOCKED = CKF_USER_PIN_LOCKED;
+        const USER_PIN_TO_BE_CHANGED = CKF_USER_PIN_TO_BE_CHANGED;
+        const SO_PIN_COUNT_LOW = CKF_SO_PIN_COUNT_LOW;
+        const SO_PIN_FINAL_TRY = CKF_SO_PIN_FINAL_TRY;
+        const SO_PIN_LOCKED = CKF_SO_PIN_LOCKED;
+        const SO_PIN_TO_BE_CHANGED = CKF_SO_PIN_TO_BE_CHANGED;
+        const ERROR_STATE = CKF_ERROR_STATE;
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+/// A limiting value for the token that may or may not take an explicit value
+pub enum Limit {
+    /// There is an explict value for this limit
+    Max(u64),
+    /// The token does not provide information about this limit
+    Unavailable,
+    /// The limit is "effectively infinite" and may be treated as such
+    Infinite,
+}
+
+/// Information about a token
+#[derive(Debug, Clone)]
+pub struct TokenInfo {
+    label: String,           // 32
```

Comment:
It'd be nice to expand the comment to say sth like "Up to/exactly 32 characters, space-padded".